pkg/dispers/conceptindexor.go

```diff
@@ -0,0 +1,200 @@
+func addSalt(concept string) error {
+
+	ConceptDoc := &ConceptDoc{
+		ConceptID:  "",
+		ConceptRev: "",
+		Concept:    concept,
+		Salt:       string(crypto.GenerateRandomBytes(defaultLenSalt)),
+	}
+	return couchdb.CreateDoc(prefixerCI, ConceptDoc)
+}
+
+func getSalt(concept string) (string, error) {
+
+	// Create index to run a mango query
+	salt := ""
+	err := couchdb.DefineIndex(prefixerCI, mango.IndexOnFields(doctypeSalt, "concept-index", []string{"concept"}))
+	if err != nil {
+		return salt, err
+	}
+
+	// Precise and run the mango query
+	var out []ConceptDoc
+	req := &couchdb.FindRequest{Selector: mango.Equal("concept", concept)}
+	err = couchdb.FindDocs(prefixerCI, doctypeSalt, req, &out)
+	if err != nil {
+		return salt, err
+	}
+
+	// Create a salt if no salt in the database for this concept
+	if len(out) == 1 {
+		salt = out[0].Salt
+	} else if len(out) == 0 {
+		return "", errors.New("Concept Indexor : no existing salt for this concept")
+	} else {
+		return "", errors.New("Concept Indexor : several salts for this concept")
+	}
+
+	return salt, err
+}
+
+func hash(str string, salt string) (string, error) {
+
+	// scrypt the cleartext string with the same parameters
+	other, err := scrypt.Key([]byte(str), []byte(salt), defaultN, defaultR, defaultP, defaultDkLen)
+	if err != nil {
+		return "", err
+	}
+	return string(other), err
+}
+
+// isConceptExisting is not useful to create a concept
+// isConceptExisting is useful to test CreateConcept or DeleteConcept
+func isConceptExisting(concept string) (bool, error) {
+
+	err := couchdb.DefineIndex(prefixerCI, mango.IndexOnFields(doctypeSalt, "concept-index", []string{"concept"}))
+	if err != nil {
+		return false, err
+	}
+
+	var out []ConceptDoc
+	req := &couchdb.FindRequest{Selector: mango.Equal("concept", concept)}
+	err = couchdb.FindDocs(prefixerCI, doctypeSalt, req, &out)
+	if err != nil {
+		return false, err
+	}
+
+	if len(out) > 0 {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+// CreateConcept is used to get a concept's salt. If the salt is absent from CI database
+// the function creates the salt and notify the user that the salt has just been created
+func CreateConcept(in *dispers.Concept) error {
+	couchdb.EnsureDBExist(prefixerCI, doctypeSalt)
+
+	// Get salt with concept
+	isExisting, err := isConceptExisting(in.Concept)
+	if err != nil {
+		return err
+	}
+
+	if !isExisting {
+		if err = addSalt(in.Concept); err != nil {
+			return err
+		}
+	}
+
+	salt, err := getSalt(in.Concept)
+	if err != nil {
+		return err
+	}
+
+	// Merge concept and salt, then hash
+	in.Hash, err = hash(in.Concept, salt)
+	return err
+}
+
+// DecryptConcept has to be used before CreateConcept or DeleteConcept
+func DecryptConcept(in *dispers.Concept) error {
+	// TODO: Decrypte concept with private key
+	in.Concept = string(in.EncryptedConcept)
+	return nil
+}
+
+// DeleteConcept is used to delete a salt in ConceptIndexor Database.
+// It has to be used to update a salt.
```

Comment:
Why would we want to update a salt?

Reply:
To change the hash associated to a concept, once a year ?

Reply:
To the best of my knowledge, there is not security reason to update hashes, but I'm open to change my mind if you have convincing resource :)